channel.js

```diff
@@ -1084,4 +1096,30 @@ TChannel.prototype.isUnhealthyError = function isUnhealthyError(err) {
     return errors.isUnhealthy(codeName);
 };
 
+TChannel.prototype.currentRetryRatio = function currentRetryPercent() {
+    if (!this.requestRateCounter) {
+        return 0;
+    }
+    return 1.0 * this.retryRateCounter / this.requestRateCounter;
+};
+
+TChannel.prototype.incrementRequest = function incrementRequest(isRetry) {
+    var self = this;
+
+    if (isRetry) {
+        incAndScheduleDec('retryRateCounter');
+    }
+    incAndScheduleDec('requestRateCounter');
+
+    function incAndScheduleDec(counterName) {
+        if (typeof self[counterName] !== 'number') {
+            return;
+        }
+        self[counterName]++;
+        self.timers.setTimeout(function offset() {
+            self[counterName]--;
```

Comment:
We can just use a different data structure like a rolling bucket, for example: https://github.com/uber/hyperbahn/blob/master/rate_limiter.js#L36-L74